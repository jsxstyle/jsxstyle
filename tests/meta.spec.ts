import { getPackages } from '@lerna/project';
import fs = require('fs');
import packlist = require('npm-packlist');
import path = require('path');
import glob = require('glob');
import * as child_process from 'child_process';

const execAsync = async (
  command: string,
  options: child_process.ExecOptions
) => {
  return new Promise<true>((resolve, reject) => {
    console.info('Running `%s` in `%s`', command, options.cwd);
    child_process.exec(command, options, (err, _stdout, _stderr) => {
      if (err) {
        reject(err);
      } else {
        resolve(true);
      }
    });
  });
};

// NOTE: this interface is incomplete
// See: @lerna/package
interface Package {
  name: string;
  location: string;
  private: boolean;
  toJSON: () => string;
}

const JSXSTYLE_ROOT = path.resolve(__dirname, '..');

// https://github.com/npm/npm-packlist/commit/63d1e3ee9c2e23ac87496ca78d3183f0652c531c
const legacySort = (a: string, b: string) =>
  // extname, then basename, then locale alphabetically
  a === 'package.json'
    ? -1
    : b === 'package.json'
    ? 1
    : /^node_modules/.test(a) && !/^node_modules/.test(b)
    ? 1
    : /^node_modules/.test(b) && !/^node_modules/.test(a)
    ? -1
    : path.dirname(a) === '.' && path.dirname(b) !== '.'
    ? -1
    : path.dirname(b) === '.' && path.dirname(a) !== '.'
    ? 1
    : a.localeCompare(b);

describe('npm publish', () => {
  it('only publishes the intended files', async () => {
    const packages: Package[] = await getPackages(JSXSTYLE_ROOT);
    const packagePromises = packages
      // exclude private packages
      .filter(pkg => !pkg.private)
      .sort((a, b) => a.name.localeCompare(b.name))
      .map(pkg =>
        // fetch file list and format it into something
        packlist({ path: pkg.location }).then(
          fileList => `
${pkg.name}
${pkg.name.replace(/./g, '=')}
${fileList
  .sort(legacySort)
  .map(f => `- ${f}`)
  .join('\n')}
`
        )
      );

    await expect(Promise.all(packagePromises)).resolves.toMatchSnapshot();
  });
});

describe('yarn.lock', () => {
  it('does not contain Twitter-internal URLs', async () => {
    const lockfileContents = fs.readFileSync('../yarn.lock', 'utf8');
    expect(lockfileContents.includes('twitter.biz')).toEqual(false);
  });
});

describe('examples', () => {
  const exampleDir = path.resolve(__dirname, '..', 'examples');
  const examples = glob.sync('jsxstyle-*-example', { cwd: exampleDir });

  for (const example of examples) {
    it(`\`${example}\` builds correctly`, async () => {
      const cwd = path.join(exampleDir, example);
      expect.assertions(1);
      return expect(execAsync('yarn build', { cwd })).resolves.toEqual(true);
    }, 30000);
  }
});
