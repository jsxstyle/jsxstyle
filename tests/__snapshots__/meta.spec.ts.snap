// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`npm publish only publishes the intended files 1`] = `
Array [
  "
gatsby-plugin-jsxstyle
======================
- package.json
- gatsby-node.js
- gatsby-ssr.js
- index.js
- LICENSE
",
  "
jsxstyle
========
- package.json
- LICENSE
- README.md
- lib/jsxstyle.cjs.js
- lib/jsxstyle.d.ts
- lib/jsxstyle.d.ts.map
- lib/jsxstyle.es.js
- lib/jsxstyle.js
- preact/lib/jsxstyle-preact.cjs.js
- preact/lib/jsxstyle-preact.d.ts
- preact/lib/jsxstyle-preact.d.ts.map
- preact/lib/jsxstyle-preact.es.js
- preact/lib/jsxstyle-preact.js
- preact/package.json
",
  "
jsxstyle-utils
==============
- package.json
- LICENSE
- README.md
- lib/addStyleToHead.d.ts
- lib/addStyleToHead.d.ts.map
- lib/componentStyles.d.ts
- lib/componentStyles.d.ts.map
- lib/dangerousStyleValue.d.ts
- lib/dangerousStyleValue.d.ts.map
- lib/getStyleCache.d.ts
- lib/getStyleCache.d.ts.map
- lib/getStyleKeysForProps.d.ts
- lib/getStyleKeysForProps.d.ts.map
- lib/hyphenateStyleName.d.ts
- lib/hyphenateStyleName.d.ts.map
- lib/jsxstyle-utils.cjs.js
- lib/jsxstyle-utils.d.ts
- lib/jsxstyle-utils.d.ts.map
- lib/jsxstyle-utils.es.js
- lib/jsxstyle-utils.js
- lib/stringHash.d.ts
- lib/stringHash.d.ts.map
- lib/types.d.ts
- lib/types.d.ts.map
",
  "
jsxstyle-webpack-plugin
=======================
- package.json
- LICENSE
- README.md
- lib/loader.d.ts
- lib/loader.d.ts.map
- lib/loader.js
- lib/loader.js.map
- lib/loaderSchema.d.ts
- lib/loaderSchema.json
- lib/plugin.d.ts
- lib/plugin.d.ts.map
- lib/plugin.js
- lib/plugin.js.map
- lib/utils/ast/accessSafe.d.ts
- lib/utils/ast/accessSafe.d.ts.map
- lib/utils/ast/accessSafe.js
- lib/utils/ast/accessSafe.js.map
- lib/utils/ast/evaluateAstNode.d.ts
- lib/utils/ast/evaluateAstNode.d.ts.map
- lib/utils/ast/evaluateAstNode.js
- lib/utils/ast/evaluateAstNode.js.map
- lib/utils/ast/extractStaticTernaries.d.ts
- lib/utils/ast/extractStaticTernaries.d.ts.map
- lib/utils/ast/extractStaticTernaries.js
- lib/utils/ast/extractStaticTernaries.js.map
- lib/utils/ast/extractStyles.d.ts
- lib/utils/ast/extractStyles.d.ts.map
- lib/utils/ast/extractStyles.js
- lib/utils/ast/extractStyles.js.map
- lib/utils/ast/generatedUid.d.ts
- lib/utils/ast/generatedUid.d.ts.map
- lib/utils/ast/generatedUid.js
- lib/utils/ast/generatedUid.js.map
- lib/utils/ast/getPropValueFromAttributes.d.ts
- lib/utils/ast/getPropValueFromAttributes.d.ts.map
- lib/utils/ast/getPropValueFromAttributes.js
- lib/utils/ast/getPropValueFromAttributes.js.map
- lib/utils/ast/getSourceModule.d.ts
- lib/utils/ast/getSourceModule.d.ts.map
- lib/utils/ast/getSourceModule.js
- lib/utils/ast/getSourceModule.js.map
- lib/utils/ast/getStaticBindingsForScope.d.ts
- lib/utils/ast/getStaticBindingsForScope.d.ts.map
- lib/utils/ast/getStaticBindingsForScope.js
- lib/utils/ast/getStaticBindingsForScope.js.map
- lib/utils/ast/parse.d.ts
- lib/utils/ast/parse.d.ts.map
- lib/utils/ast/parse.js
- lib/utils/ast/parse.js.map
- lib/utils/getClassNameFromCache.d.ts
- lib/utils/getClassNameFromCache.d.ts.map
- lib/utils/getClassNameFromCache.js
- lib/utils/getClassNameFromCache.js.map
- lib/utils/getStylesByClassName.d.ts
- lib/utils/getStylesByClassName.d.ts.map
- lib/utils/getStylesByClassName.js
- lib/utils/getStylesByClassName.js.map
- lib/utils/wrapFileSystem.d.ts
- lib/utils/wrapFileSystem.d.ts.map
- lib/utils/wrapFileSystem.js
- lib/utils/wrapFileSystem.js.map
",
]
`;
